mlshvch/hometask-3/rainfall.rb

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+file = File.open('data.txt', 'r')
+data = file.read
+file.close
+
+EXIT_COMMAND = 'exit!'
+ERROR_MESSAGE = 'City name can not be blank!'
+DEFAULT_VALUE = -1.0
+
+def extract_data(data_string)
+  data_string.split("\n").each_with_object({}) do |data, city_rainfall_stats|
+    town, rainfall_stats = data.split(':')
+    city_rainfall_stats[town] = rainfall_stats.split(',').each_with_object([]) do |results, month_rainfall_stats|
+      month_rainfall_stats << results.split(' ')[1]
+    end
+  end
+end
+
+@rainfall_data = extract_data(data)
+
+def mean(town, data)
+  return DEFAULT_VALUE unless data[town]
+
+  data[town].sum(&:to_f) / 12
+end
+
+def variance(town, data)
+  return DEFAULT_VALUE unless data[town]
+
+  mean_value = mean(town, @rainfall_data)
+  data[town].sum { |value| (value.to_f - mean_value)**2 } / 12
+end
+
+def run_cli
+  loop do
+    puts 'Enter city name:'
+    input = gets.chomp
+    break if input == EXIT_COMMAND
+
+    if input.empty?
+      puts ERROR_MESSAGE
+    else
+      puts "Rainfall mean: #{mean(input, @rainfall_data)}"
```

Comment:
Better than it was before, but still not the best approach. You see, you've just created an instance variable `@rainfall_data`, but you don't have any other enclosing context to isolate it. Your Ruby code is currently written in procedural style, which means that it is executed in a scope of `main` object - the topmost object, where all the other defined methods, constants and the rest reside. Since it's a global topmost object, any instance variable will live there until the execution of Ruby process stops and can possibly override any other existing instance variables with the same name in `main` object itself. In other words, this approach has the almost the same downsides as using global variables and highly discouraged to be used in `main` scopes. If you don't have any modules or classes to which this instance variable could belong, then it's better to not use it.

The whole lifecycle of your program is within one method, `run_cli`. What's better to do here is to read the contents once before the beginning of the `loop` statement and then just pass the local to `mean` / `variance` the usual way.